feat: Prefab QoL [MTT-5116]

ShadauxCat opened the pull request:
Moves prefab list to a ScriptableObject and makes it publicly accessible to be programmable. Credit for most of the work goes to @JesseOlmer, I provided some refinements and testing (including unit tests)

## Changelog

- Changed: Network prefabs are now stored in a ScriptableObject that can be shared between NetworkManagers, and have been exposed for public access. By default, a Default Prefabs List is created that contains all NetworkObject prefabs in the project, and new NetworkManagers will default to using that unless that option is turned off in the Netcode for GameObjects settings. Existing NetworkManagers will maintain their existing lists, which can be migrated to the new format via a button in their inspector.

## Testing and Documentation

- Includes unit tests.
- Includes documentation for previously-undocumented public API entry points.

ShadauxCat committed:
feat: Prefab QoL

Moves prefab list to a ScriptableObject and makes it publicly accessible to be programmable. Credit for most of the work goes to @JesseOlmer, I provided some refinements and testing (including unit tests)

ShadauxCat committed:
standards

ShadauxCat committed:
Merge branch 'develop' into feat/prefab_qol

ShadauxCat committed:
Standards again.

ShadauxCat committed:
Changelog

ShadauxCat committed:
compile fix

ShadauxCat committed:
Another compile fix (tools-integration)

becksebenius-unity commented on com.unity.netcode.gameobjects/Editor/Configuration/NetcodeForGameObjectsSettings.cs:
```diff
@@ -35,5 +43,39 @@ internal static void SetAutoAddNetworkObjectSetting(bool autoAddSetting)
             EditorPrefs.SetBool(AutoAddNetworkObjectIfNoneExists, autoAddSetting);
         }
+
+        public static NetcodeForGameObjectsSettings GetOrCreateSettings()
```
**Comment:** It seems a little odd to me that this class is a mixture of editor prefs and persistent prefs.
**Suggestion:** If you want to stay consistent with Unity's asset serialization and also be able to take advantage of things like `[FormerlySerializedAs]` you can use these APIs to read/write scriptable object assets to the project folder:

```
UnityEditorInternal.InternalEditorUtility.SaveToSerializedFileAndForget();
UnityEditorInternal.InternalEditorUtility.LoadSerializedFileAndForget();
```

Or alternatively, you can use `ScriptableSingleton` which does most of that for you automatically.

ShadauxCat replied:
I think I see what you're saying... but this doesn't actually need to be persistent prefs, it can use EditorPrefs everywhere. The mismatch was the result of merging two different implementations of this together and not noticing that they were implemented in different ways.

JesseOlmer replied:
I think they actually need to be separate - the existing editorprefs are really editor preferences. how should this editor on this PC behave. The prefab prefs are project-level... you want that behavior consistent on everyone's PC who is working on the project.

ShadauxCat replied:
Ah, fair. I didn't realize EditorPrefs were shared between projects.

NoelStephensUnity replied:
I had the same knee jerk reaction when looking at the original implementation, but there are advantages to having the NetworkPrefabs configuration serialized in reasonably "readable" json formatted text file.  This could be easily converted over to a editor preferences via [EditorPrefs.SetString](https://docs.unity3d.com/ScriptReference/EditorPrefs.SetString.html), but I am not sure a netcode project's NetworkPrefab settings (specific to the game/product) qualifies as an editor preference.

Might be worth investigating the ScriptableObject path... just to determine if we are gaining or losing potential future functionality (either side of the fence).

The good news is that the json text can be easily saved to and read from whichever medium/location turns out to be the best place to store this information.  👍

ShadauxCat replied:
Switched it over to ScriptableSingleton. That eased some of the logic and it's still a plain text format. :)

becksebenius-unity commented on com.unity.netcode.gameobjects/Editor/Configuration/NetworkPrefabProcessor.cs:
```diff
@@ -0,0 +1,160 @@
+        }
+        private static NetworkPrefabsList s_PrefabsList;
+
+        // Unfortunately this method is required by the asset pipeline to be static
```
**Nitpick:** This comment seems extraneous, it doesn't explain what's unfortunate about this pattern or why it shouldn't be static. As an engineer without context about this code, I think this actually detracts from it rather than improves it

ShadauxCat replied:
That comment came from @JesseOlmer so I won't speak for him as to the intention behind it.

JesseOlmer replied:
```suggestion
```

This was mostly a note to myself during experimentation... It's unfortunate because static things are harder to mock, override, disable, etc. I didn't actually write ANY tests for any of this code so not sure how much this being static actually made things harder ;)

ShadauxCat replied:
Had a feeling that was your thought process here... and yeah, it made testing harder... tests are having to set static variables and make sure they unset them at the end because if they don't, the actual editor settings get changed... it's yucky.

NoelStephensUnity replied:
Yeah...it is what it is:
https://docs.unity3d.com/2022.2/Documentation/ScriptReference/AssetPostprocessor.OnPostprocessAllAssets.html
One of those scenarios where creating an internal set of callbacks might help (?) with testing (and be valid due to it being static and all).

becksebenius-unity commented on com.unity.netcode.gameobjects/Editor/NetworkManagerEditor.cs:
```diff
@@ -224,7 +157,26 @@ public override void OnInspectorGUI()
+                {
+                    if (m_NetworkManager.NetworkConfig.Prefabs.NetworkPrefabsList == null)
+                    {
+                        EditorGUILayout.HelpBox("You have no prefab list selected. You will have to add your prefabs manually at runtime for netcode to work.", MessageType.Warning);
```
**Question:** Should this be `Info`? It seems like having no prefabs is a valid use case.

ShadauxCat replied:
It is a valid use case, but not an easy or intuitive one. People have to write their own code to manage the prefabs list on both sides. I felt warning was more appropriate to call out "this isn't going to work unless you're really putting effort into making it work"... but not opposed to changing it to info

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkPrefab.cs:
```diff
@@ -3,7 +3,7 @@
 namespace Unity.Netcode
 {
-    internal enum NetworkPrefabOverride
+    public enum NetworkPrefabOverride
```
**Comment:** I think this needs xmldocs

ShadauxCat replied:
to be completely honest, I've never actually understood the difference between these values. I'll see if I can find someone to explain it so I can write an xmldoc...

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkPrefabs.cs:
```diff
@@ -0,0 +1,283 @@
+namespace Unity.Netcode
+{
+    [Serializable]
+    public class NetworkPrefabs
```
**Comment:** I think this needs xml docs

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkPrefabs.cs:
```diff
@@ -0,0 +1,283 @@
+        }
+
+        /// <summary>
+        /// Processes the <see cref="NetworkPrefabsList"/> if one is present for use during runtime execution,
```
**Nitpick:** one what? I find this documentation confusing

ShadauxCat replied:
I think it might be the xml notation that's confusing you.

"Processes the NetworkPrefabsList if one is present"

To me that makes sense ("Processes the NetworkPrefabsList if a NetworkPrefabsList is present")

becksebenius-unity replied:
Oh, I think you're right.

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkPrefabs.cs:
```diff
@@ -0,0 +1,283 @@
+            }
+        }
+
+        public bool Contains(GameObject prefab)
```
**Comment:** Missing xmldoc (other overload too)

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Configuration/NetworkPrefabsList.cs:
```diff
@@ -0,0 +1,78 @@
+namespace Unity.Netcode
+{
+    [CreateAssetMenu(fileName = "NetworkPrefabsList", menuName = "Netcode/Network Prefabs List")]
+    public class NetworkPrefabsList : ScriptableObject
```
**Comment:** Missing xmldoc (for some of the methods too)

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -508,6 +508,19 @@ public struct ConnectionApprovalRequest
+        internal static ResetNetworkManagerDelegate OnNetworkManagerReset;
+#endif
+
+        private void Reset()
```
**Question:** Was this meant to be invoked from somewhere? I don't see any calls to it.

ShadauxCat replied:
This is a Unity event like Awake, Update, etc. It's called when the component is first created/added to the GameObject (in the editor), as well as when the user hits "reset" on the component in the inspector.

I'm unsure if it's ever called at runtime or not.

becksebenius-unity replied:
Oh, I totally forgot about that! Thanks :)

NoelStephensUnity commented on com.unity.netcode.gameobjects/Editor/Configuration/NetcodeSettingsProvider.cs:
```diff
@@ -40,6 +43,32 @@ private static void OnGuiHandler(string obj)
                 NetcodeForGameObjectsSettings.SetAutoAddNetworkObjectSetting(autoAddNetworkObjectSetting);
                 NetcodeForGameObjectsSettings.SetNetcodeInstallMultiplayerToolTips(multiplayerToolsTipStatus ? 0 : 1);
             }
+            const string generateNetworkPrefabsString = "Generate Default Network Prefabs List";
```
You might think about migrating all of the prefab UI settings stuff within the same EditorGUI.BeginChangeCheck(); and EditorGUI.EndChangeCheck() region.  
All of the initialization inserted between lines 35 and 36.
Lines 60-66 inserted between 40-41.
Lines 69-70 inserted after line 44.

ShadauxCat replied:
Reworked this whole section. The display is nicer now. Got little collapsies and everything.

![image](https://user-images.githubusercontent.com/284434/203646031-e0c26f38-728d-486d-b7a6-f251dd527828.png)

NoelStephensUnity replied:
😻 ❤️‍🔥

ShadauxCat committed:
review feedback

ShadauxCat committed:
Style + reverted some testing code that wasn't used.

ShadauxCat committed:
removed the old settings file, it's got a new name now

ShadauxCat committed:
Review feedback:

- When performing a migration from the old prefab list to the new one, the path used to generate the new prefab list will be:
  - Next to the prefab containing the NetworkManager component if the NetworkManager is part of a prefab
  - Next to the scene the NetworkManager is in if it's not part of a prefab

- NetworkManager can now have multiple prefab lists attached to it

Also, I moved the code from NetworkAnimator's OnBeforeSerialize into the OnValidate event, as OnBeforeSerialize gets called in various circumstances that result in Unity throwing an error now that a NetworkAnimator can be a part of a NetworkPrefabsList.

ShadauxCat committed:
Revert NetworkAnimator change

ShadauxCat committed:
Conditional compilation for API change in 2021

ShadauxCat committed:
Merge branch 'develop' into feat/prefab_qol

# Conflicts:
#	com.unity.netcode.gameobjects/CHANGELOG.md
#	com.unity.netcode.gameobjects/Editor/Configuration/NetcodeSettingsProvider.cs

ShadauxCat committed:
Standards

NoelStephensUnity reviewed (approved):
Looks good Kitty!  👍
